File: modules/vescape-core/android/src/main/java/expo/modules/vescapecore/accessory/BrakeLight.kt
package expo.modules.vescapecore.accessory

import kotlin.math.abs

/** Speed is km/h, clock is monotonic. PoC defaults documented in docs/accessories.md.
 * @parity /modules/vescape-core/ios/accessory/BrakeLight.swift
 */
class BrakeLightDetector {
    private var previousSpeed: Double? = null
    private var previousAt: Long? = null
    private var deceleration = 0.0
    var mode: String? = null
        private set

    fun clear() { previousSpeed = null; previousAt = null; deceleration = 0.0; mode = null }

    fun sample(speedKmh: Double, riding: Boolean, at: Long, sensitivity: Int) {
        if (!speedKmh.isFinite()) { clear(); return }
        val speed = abs(speedKmh) / 3.6
        val oldSpeed = previousSpeed
        val dt = previousAt?.let { at - it }
        previousSpeed = speed
        previousAt = at
        if (!riding) { deceleration = 0.0; mode = "not_riding"; return }
        if (oldSpeed == null || dt == null) { mode = "riding"; return }
        if (dt <= 0 || dt > 500) { deceleration = 0.0; mode = null; return }
        val seconds = dt / 1000.0
        val alpha = seconds / (0.2 + seconds)
        deceleration += alpha * ((oldSpeed - speed) / seconds - deceleration)
        val scale = 1.5 - sensitivity / 100.0
        val braking = scale * if (mode == "braking") 0.75 else 1.0
        val hard = scale * if (mode == "hard_braking") 2.25 else 3.0
        mode = when { deceleration >= hard -> "hard_braking"; deceleration >= braking -> "braking"; else -> "riding" }
    }
}

/** Persisted per Accessory capability; automatic state follows whichever Board is current.
 * @parity /modules/vescape-core/ios/accessory/BrakeLight.swift `BrakeLightSettings`
 * @parity /modules/vescape-core/src/index.ts `BrakeLightSettings`
 */
data class BrakeLightSettings(val sensitivity: Int = 50, val parked: String = "off") {
    fun valid() = sensitivity in 1..100 && parked in setOf("off", "glow")
    fun toMap(): Map<String, Any?> = mapOf("sensitivity" to sensitivity, "parked" to parked)
}

/** @parity /modules/vescape-core/ios/accessory/BrakeLight.swift `BrakeLightController` */
class BrakeLightController {
    data class Key(val accessoryId: String, val capabilityId: String)
    private class Light(var settings: BrakeLightSettings = BrakeLightSettings()) {
        val detector = BrakeLightDetector()
        var preview: String? = null
    }
    private val lights = linkedMapOf<Key, Light>()
    private var riding = false
    fun configure(key: Key, settings: BrakeLightSettings) { lights.getOrPut(key) { Light() }.settings = settings }
    fun forget(accessoryId: String) { lights.keys.removeAll { it.accessoryId == accessoryId } }
    fun sample(speed: Double, engaged: Boolean, at: Long) {
        riding = engaged
        lights.values.forEach { if (engaged) it.preview = null; it.detector.sample(speed, engaged, at, it.settings.sensitivity) }
    }
    fun clear() { riding = false; lights.values.forEach { it.detector.clear() } }
    fun releasePreviews() { lights.values.forEach { it.preview = null } }
    fun preview(key: Key, mode: String?): Boolean {
        if (mode != null && (riding || mode !in MODES)) return false
        val light = lights[key] ?: return false
        light.preview = mode
        return true
    }
    /** @parity /modules/vescape-core/src/index.ts `AccessoryCapability` */
    fun describe(key: Key): Map<String, Any?> = lights.getOrPut(key) { Light() }.let {
        mapOf("brakeLight" to it.settings.toMap(), "lightMode" to it.detector.mode, "lightPreview" to it.preview)
    }
    fun command(key: Key): AccessoryCommand.State = lights.getOrPut(key) { Light() }.let {
        AccessoryCommand.State(key.capabilityId, if(it.detector.mode == null) "unavailable" else "available", it.preview ?: it.detector.mode, it.settings.parked, it.preview != null)
    }
    companion object { val MODES = setOf("riding", "braking", "hard_braking", "not_riding") }
}

File: modules/vescape-core/ios/accessory/BrakeLight.swift
import Foundation

/// @parity /modules/vescape-core/android/src/main/java/expo/modules/vescapecore/accessory/BrakeLight.kt
final class BrakeLightDetector {
  private var previousSpeed: Double?
  private var previousAt: Int64?
  private var deceleration = 0.0
  private(set) var mode: String?
  func clear() {
    previousSpeed = nil
    previousAt = nil
    deceleration = 0
    mode = nil
  }
  func sample(speedKmh: Double, riding: Bool, at: Int64, sensitivity: Int) {
    guard speedKmh.isFinite else {
      clear()
      return
    }
    let speed = abs(speedKmh) / 3.6
    let oldSpeed = previousSpeed
    let dt = previousAt.map { at - $0 }
    previousSpeed = speed
    previousAt = at
    guard riding else {
      deceleration = 0
      mode = "not_riding"
      return
    }
    guard let oldSpeed, let dt else {
      mode = "riding"
      return
    }
    guard dt > 0 && dt <= 500 else {
      deceleration = 0
      mode = nil
      return
    }
    let seconds = Double(dt) / 1000
    let alpha = seconds / (0.2 + seconds)
    deceleration += alpha * ((oldSpeed - speed) / seconds - deceleration)
    let scale = 1.5 - Double(sensitivity) / 100
    let braking = scale * (mode == "braking" ? 0.75 : 1)
    let hard = scale * (mode == "hard_braking" ? 2.25 : 3)
    mode = deceleration >= hard ? "hard_braking" : deceleration >= braking ? "braking" : "riding"
  }
}

/// @parity /modules/vescape-core/android/src/main/java/expo/modules/vescapecore/accessory/BrakeLight.kt `BrakeLightSettings`
/// @parity /modules/vescape-core/src/index.ts `BrakeLightSettings`
struct BrakeLightSettings {
  var sensitivity = 50
  var parked = "off"
  var valid: Bool { (1...100).contains(sensitivity) && ["off", "glow"].contains(parked) }
  func toMap() -> [String: Any?] { ["sensitivity": sensitivity, "parked": parked] }
}

/// @parity /modules/vescape-core/android/src/main/java/expo/modules/vescapecore/accessory/BrakeLight.kt `BrakeLightController`
final class BrakeLightController {
  struct Key: Hashable {
    let accessoryId: String
    let capabilityId: String
  }
  private final class Light {
    var settings = BrakeLightSettings()
    let detector = BrakeLightDetector()
    var preview: String?
  }
  private var lights: [Key: Light] = [:]
  private var riding = false
  private func light(_ key: Key) -> Light {
    if let light = lights[key] { return light }
    let light = Light()
    lights[key] = light
    return light
  }
  func configure(_ key: Key, _ settings: BrakeLightSettings) { light(key).settings = settings }
  func forget(_ accessoryId: String) {
    lights = lights.filter { $0.key.accessoryId != accessoryId }
  }
  /// Returns whether anything a screen renders changed, so an unchanged sample publishes nothing.
  @discardableResult
  func sample(speed: Double, engaged: Bool, at: Int64) -> Bool {
    riding = engaged
    var changed = false
    for light in lights.values {
      // Riding ends a preview: the rider is on the board and the light follows the board.
      if engaged, light.preview != nil {
        light.preview = nil
        changed = true
      }
      let before = light.detector.mode
      light.detector.sample(
        speedKmh: speed, riding: engaged, at: at, sensitivity: light.settings.sensitivity)
      if light.detector.mode != before { changed = true }
    }
    return changed
  }
  func clear() {
    riding = false
    lights.values.forEach { $0.detector.clear() }
  }
  func releasePreviews() { lights.values.forEach { $0.preview = nil } }
  func preview(_ key: Key, mode: String?) -> Bool {
    if let mode, riding || !Self.modes.contains(mode) { return false }
    guard let light = lights[key] else { return false }
    light.preview = mode
    return true
  }
  /// @parity /modules/vescape-core/src/index.ts `AccessoryCapability`
  func describe(_ key: Key) -> [String: Any?] {
    let light = light(key)
    return [
      "brakeLight": light.settings.toMap(), "lightMode": light.detector.mode,
      "lightPreview": light.preview,
    ]
  }
  func command(_ key: Key, enabled: Bool = true) -> AccessoryCommand {
    let light = light(key)
    if !enabled { return .state(capabilityId: key.capabilityId, telemetry: "available", mode: "not_riding", parked: "off", preview: false) }
    return .state(
      capabilityId: key.capabilityId,
      telemetry: light.detector.mode == nil ? "unavailable" : "available",
      mode: light.preview ?? light.detector.mode, parked: light.settings.parked,
      preview: light.preview != nil)
  }
  static let modes = ["riding", "braking", "hard_braking", "not_riding"]
}
